homeassistant/components/sensor/paloalto.py

```diff
@@ -0,0 +1,266 @@
+        return self._api.data[self._var_id]
+
+    @property
+    def available(self):
+        """Could the device be accessed during the last update call."""
+        return self._api.available
+
+    def update(self):
+        """Get the latest data and updates the state."""
+        self._api.update()
+
+
+class PaloAltoApi(object):
+    """The class for handling the data retrieval from Palo Alto Device."""
+
+    def __init__(self, host, use_ssl, verify_ssl, api_key):
+        """Initialize the Palo Alto API."""
+        self._host = host
+        self._use_ssl = use_ssl
+        self._verify_ssl = verify_ssl
+        self._api_key = api_key
+        self._usersdata = None
+        self._sysinfo = None
+        self._gp_users = None
+        self._temperature = None
+        self.available = True
+        self._sensors = {}
+
+    @property
+    def data(self):
+        """Return data."""
+        return self._sensors
+
+    def get_uri_scheme(self, use_ssl):
+        """Return proper uril scheme based on config setting."""
+        return 'https://' if use_ssl else 'http://'
+
+    def get_resource(self, use_ssl, host, api_key, endpoint):
+        """Prepare the URL."""
+        uri_scheme = self.get_uri_scheme(use_ssl)
+        if endpoint == EndPointType.Operational:
+            return "{}{}{}&key={}".format(uri_scheme, self._host,
+                                          CONST_OPS_ENDPOINT, self._api_key)
+        else:
+            return "{}{}{}&key={}".format(uri_scheme, self._host,
+                                          CONST_CONFIG_ENDPOINT, self._api_key)
+
+    def http_request(self, url):
+        """HTTP request to the Palo Alto device."""
+        content = None
+        context = None
+        try:
+            if self._use_ssl and not self._verify_ssl:
+                context = ssl._create_unverified_context()
+            response = urllib.request.urlopen(url, context=context)
+            content = response.read()
+        except Exception as ex:
+            _LOGGER.error(str(ex))
+            content = None
+
+        return content
+
+    def update(self):
+        """Get Operational and Configuration urls."""
+        ops_url = self.get_resource(self._use_ssl, self._host,
+                                    self._api_key, EndPointType.Operational)
+
+        users_url = ops_url.replace(CONST_COMMAND, PA_OPS_ACTIVE_USERS)
+        self._usersdata = self.http_request(users_url)
+
+        sysinfo_url = ops_url.replace(CONST_COMMAND, PA_CONF_SYS_INFO)
+        self._sysinfo = self.http_request(sysinfo_url)
+
+        gp_users_url = ops_url.replace(CONST_COMMAND, PA_CONF_GP_USERS)
+        self._gp_users = self.http_request(gp_users_url)
+
+        temperature_url = ops_url.replace(CONST_COMMAND, PA_CONF_TEMPERATURE)
+        self._temperature = self.http_request(temperature_url)
+
+        """parse the xml data"""
+        self.parse_data()
+
+    def parse_globalprotect_users(self):
```

Comment:
All this parsing code has to be extracted into an external lib which is then used by Home Assistant.

Reply:
That totally makes sense, Where should I place that library? Are there any references you can provide?

Reply:
He means PyPi. You should place the device specific code in a separate lib on PyPi.

Reply:
@balloob Can you please confirm? I'd rather get a confirmation before I go down any path at this time. It has been a painful learning, I'd rather not aggravate it 😃

Reply:
@arsaboo is right.

Reply:
@balloob thanks for confirming. I wish these things are documented somewhere on the developer portal. It would save a lot of time for everyone. Also, there were a few comments above - one about `unique_id` and another about your statement on "you are not allowed to have changing times in Home Assistant". To me it sounds like a best practice, and any documentation on this would be helpful as well.

Reply:
That we require external lib is in the docs: https://home-assistant.io/developers/code_review_platform/#6-communication-with-devicesservices

A template for a lib is not. I have on my to do list to make a "starter pack". For now, use this example: https://github.com/balloob/pynetgear